### PR TITLE
linux: add fallocate()

### PR DESCRIPTION
**Pull request by jorangreef**

### CODE REVIEW

File: lib/std/os/linux.zig [hunk 1]
```diff
@@ -121,6 +121,16 @@ pub fn utimensat(dirfd: i32, path: ?[*:0]const u8, times: *const [2]timespec, fl
     return syscall4(.utimensat, @bitCast(usize, @as(isize, dirfd)), @ptrToInt(path), @ptrToInt(times), flags);
 }
 
+pub fn fallocate(fd: i32, mode: i32, offset: usize, len: usize) usize {
```
Comment (daurnimator) [inline lib/std/os/linux.zig:124]: should `offset` be a `u64`?

**Review by daurnimator — COMMENTED** [review]
Need in `std.c` as well?

File: lib/std/os/linux.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/os/linux.zig:124]: > should offset be a u64?

Both `offset` and `usize` are `off_t` (so `u64`) and may need special care when passed to the kernel and be split in `u32`/`u32`.

**Comment by LemonBoy** [comment]
> Need in std.c as well?

In `c/linux.zig`, this is a linux-only API.

We also need the definitions for the `mode` flags.

A test in `os/linux/test.zig` is also needed.

File: lib/std/os/linux.zig (hunk 1 above)
Comment (jorangreef) [reply lib/std/os/linux.zig:124]: > Both offset and usize are off_t (so u64)

I assumed that `off_t` is guaranteed to be at least an `int` but is not always `u64`, which is why I went with `usize`, thinking that would be 32-bit on 32-bit machines else 64-bit. You will know better than me!

`off_t` seems complicated, according to https://stackoverflow.com/a/14351239/2138743:

> This is a signed integer type used to represent file sizes. 
> In the GNU C Library, this type is no narrower than int.
> If the source is compiled with _FILE_OFFSET_BITS == 64 this 
> type is transparently replaced by off64_t.

How would we know if the source is compiled with `_FILE_OFFSET_BITS == 64`?

Comment (jorangreef) [reply lib/std/os/linux.zig:124]: > may need special care when passed to the kernel and be split in u32/u32.

What would this special care look like?

**Comment by jorangreef** [comment]
> In c/linux.zig, this is a linux-only API.

What signature would you propose we use here?

> We also need the definitions for the mode flags.

I assume these would go in `os/bits/linux.zig`?

File: lib/std/os/linux.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/os/linux.zig:124]: > I assumed that off_t is guaranteed to be at least an int but is not always u64, which is why I went with usize, thinking that would be 32-bit on 32-bit machines else 64-bit. You will know better than me!

Don't mix up libc definitions and kernel ones, the latter is defined as `long long`.

> How would we know if the source is compiled with _FILE_OFFSET_BITS == 64?

Most of the stdlib assumes that's true, use `off_t` in the libc definition (even though that's signed...).

> What would this special care look like in more detail? I'm new to this, so if you could provide a sample that would be a huge help.

32bit kernels work on 32bit registers so passing the offset requires some care. Check out musl's `fallocate.c` wrapper for more info, I usually double-check if the values are sane using `strace`.

> What signature would you propose we use here?

Follow the libc signature, they have to match:
```zig
pub extern "c" fn fallocate(fd: fd_t, mode: c_int, offset: off_t, len: off_t) c_int
```

> I assume these would go in os/bits/linux.zig?

Yup

Comment (LemonBoy) [reply lib/std/os/linux.zig:124]: > With that in mind, would it be fine if we do a branch on @sizeOf(usize) and then select syscall4 (64-bit kernel) or syscall6 (32-bit kernel)?

Yep

> What would be the ziggest way of splitting the long long into two longs

Shift&truncate

Comment (jorangreef) [reply lib/std/os/linux.zig:124]: So musl's `fallocate.c` looks like this:

```c
int fallocate(int fd, int mode, off_t base, off_t len)
{
    return syscall(SYS_fallocate, fd, mode, __SYSCALL_LL_E(base),
        __SYSCALL_LL_E(len));
}
```

Where `__SYSCALL_LL_E` is defined as an identity function to pass one `long long` parameter, or else two `long` parameters to the variadic `syscall` function, depending on the architecture:

```c
#define __SYSCALL_LL_E(x) (x)
```

or

```c
#define __SYSCALL_LL_E(x) \
((union { long long ll; long l[2]; }){ .ll = x }).l[0], \
((union { long long ll; long l[2]; }){ .ll = x }).l[1]
```

musl defines these per architecture, using a single `long long` parameter for:

```
aarch64
mips64
powerpc64
riscv64
s390x
x86_64
```

and two `long` parameters for:

```
arm
i386
mips
powerpc
```

With that in mind, would it be fine if we do a branch on `@sizeOf(usize)` and then select `syscall4` (64-bit kernel) or `syscall6` (32-bit kernel)?

What would be the *ziggest* way of splitting the `long long` into two `longs`? `@bitCast([2]u32, offset)` assuming `offset` becomes a `u64` in our `fallocate()` function signature as @daurnimator suggested?

Comment (jorangreef) [reply lib/std/os/linux.zig]: Done

Comment (jorangreef) [reply lib/std/os/linux.zig:124]: ...I see `llseek()` simply does `@truncate(usize, offset >> 32), @truncate(usize, offset)`

File: lib/std/os/linux.zig [hunk 2]
```diff
@@ -121,6 +121,28 @@ pub fn utimensat(dirfd: i32, path: ?[*:0]const u8, times: *const [2]timespec, fl
     return syscall4(.utimensat, @bitCast(usize, @as(isize, dirfd)), @ptrToInt(path), @ptrToInt(times), flags);
 }
 
+pub fn fallocate(fd: i32, mode: i32, offset: u64, len: u64) usize {
+    if (@sizeOf(usize) == 4) {
+        return syscall6(
+            .fallocate,
+            @bitCast(usize, @as(isize, fd)),
+            @bitCast(usize, @as(isize, mode)),
+            @truncate(usize, offset >> 32),
```
Comment (LemonBoy) [inline lib/std/os/linux.zig:130]: Pro tip: try running this test with `zig test -target i386-linux ../lib/std/os/linux/test.zig`, you swapped the upper/lower halves.

Comment (jorangreef) [reply lib/std/os/linux.zig]: Thanks for the tip! Can we always assume little-endian kernel for `syscall6` all 32-bit targets?

**Comment by LemonBoy** [comment]
I can't locally reproduce the mipsel crash, perhaps it's a qemu bug?

**Comment by andrewrk** [comment]
If it's a QEMU bug, we should file a bug report upstream, because thanks to https://github.com/ziglang/qemu-static we are using the latest release (5.1.0) on the CI.

In the meantime, feel free to add code to skip the test like this:

```
if (condition in which the test is failing) {
    // TODO: <link to github issue to track the failing test case>
    return error.SkipZigTest;
}
```

**Comment by LemonBoy** [comment]
lel. It turns out somebody added the big-endian mips to the testing roster _but_ every problem was solved by skipping the failing tests (#5127), that's how you do it 'round here.

Ok, it turns out we have to audit&unfuck every syscall where file offsets are passed in... every other non-failing test is probably called with 0 as offset so the endianness doesn't matter...

+1 from me if:
* The test is skipped with `if (std.Target.current.cpu.arch == .mips) return error.SkipZigTest;`
* A link to `https://github.com/ziglang/zig/issues/5127` is added in a comment
* In case of unexpected `errno` use `std.debug.panic` instead of skipping the test

**Comment by jorangreef** [comment]
> Ok, it turns out we have to audit every syscall where file offsets are passed in... every other non-failing test is probably called with 0 as offset so the endianness doesn't matter...

I guess I'm glad we found this. In this case, big-endian mips was giving errno 27 File Too Large.

> +1 from me if:
> - The test is skipped with if (std.Target.current.cpu.arch == .mips) return error.SkipZigTest;
> - A link to https://github.com/ziglang/zig/issues/5127 is added in a comment
> - In case of unexpected errno use std.debug.panic instead of skipping the test

Thanks, done.

**Comment by andrewrk** [comment]
Now that #5127 is solved, does the test case pass?

**Comment by jorangreef** [comment]
It should, but a gimli test is failing from the related changes I think.

**Comment by jorangreef** [comment]
Could someone please trigger the CI to re-run? We should be fine now, assuming the Gimli issue that crept in to the build has been fixed.

**Comment by data-man** [comment]
@jorangreef 
> Could someone please trigger the CI to re-run?

Close and reopen PR?

**Comment by jorangreef** [comment]
Test passing on mips big-endian. Ready to land!

**Comment by andrewrk** [comment]
Thanks!

**Pull request merged**